File: chapter_0/dynamic_programming/fibonacci_sequence/fib.rs
/*
 * @Author: Mengsen.Wang
 * @Date: 2021-01-19 09:46:53
 * @Last Modified by:   Mengsen.Wang
 * @Last Modified time: 2021-01-19 09:46:53
 */

fn just_recursive(n: i32) -> i32 {
    if n == 1 || n == 2 {
        return 1;
    }

    just_recursive(n - 1) + just_recursive(n - 2)
}

fn dp_fib(n: i32) -> i32 {
    let mut dp: Vec<i32> = vec![0; n as usize + 1];
    dp[0] = 1;
    dp[1] = 1;
    for i in 3..=n {
        dp[i as usize] = dp[i as usize - 1] + dp[i as usize - 2];
    }
    dp[n as usize]
}
fn fib(n: i32) -> i32 {
    (0..n).fold((0, 1), |(pp, p), _| (p, pp + p)).0
}

fn main() {
    println!("The result = {}", just_recursive(45));
    println!("The result = {}", dp_fib(45));
    println!("The result = {}", fib(45));
}

File: cpp/fibonacci_sequence/fib.cpp
/*
 * @Author: Mengsen.Wang
 * @Date: 2021-01-19 09:46:27
 * @Last Modified by: Mengsen.Wang
 * @Last Modified time: 2021-01-19 09:46:51
 */

#include <cstdio>
#include <string>
#include <vector>

int just_recursive(int N) {
  if (N == 1 || N == 2) return 1;
  return just_recursive(N - 1) + just_recursive(N - 2);
}

int dp_fib(int N) {
  std::vector<int> dp(N + 1, 0);

  dp[1] = dp[2] = 1;
  for (int i = 3; i <= N; ++i) {
    dp[i] = dp[i - 1] + dp[i - 2];
  }
  return dp[N];
}

int fib(int n) {
  if (n == 2 || n == 1) return 1;
  int prev = 1, curr = 1, sum = 0;
  for (int i = 3; i <= n; ++i) {
    sum = prev + curr;
    prev = curr;
    curr = sum;
  }
  return curr;
}

int main() {
  printf("The result = %d\n", just_recursive(45));
  printf("The result = %d\n", dp_fib(45));
  printf("The result = %d\n", fib(45));
  return 0;
}
